### ISSUE TITLE
ERB code in pages

### ISSUE THREAD

**Issue by chalmagean** [sitepress/sitepress#34]
Hi,

I'm using a helper in the page content. Something like the following:

```
## Subhead

Some text

<%= my_helper %>
```

This doesn't work because the ERB tags are not evaluated.

```
class MarkdownTemplateHandler
  def erb
    @erb ||= ActionView::Template.registered_template_handler(:erb)
  end

  def call(template, source)
    compiled_source = erb.call(template, source)

    "Redcarpet::Markdown.new(Redcarpet::Render::HTML.new, fenced_code_blocks: true).render(begin;#{compiled_source};end).html_safe"
  end
end

ActionView::Template.register_template_handler(:md, MarkdownTemplateHandler.new)
```

**Comment by bradgessler** [comment]
Yeah, the example you show is exactly how the internals of the rails-markdown renderer works: https://github.com/sitepress/markdown-rails/blob/main/lib/markdown-rails/handler/erb.rb#L8

There's many answers for this:

**1. "You're holding it the wrong way"**

I try to avoid mixing Erb and Markdown, and can accomplish this a vast majority of the time by using frontmatter and templates to handle whatever stuff I need to render outside of markdown. 

I've seen the inline rendering of markdown introduce subtle bugs too when folks are unclear on if the MD renders first or the ERB.

As much as I'd like for this to be the right answer, its not.

**2. Actually we need an MD parser that exposes an AST in Ruby**

The way I define this issue is, "don't evaluate ERB inside of code fences or code blocks". The best way to accomplish that is for the MD render to expose an abstract syntax tree (AST) and have erb evaluate appropriate blocks, but ignore the others. The closest Ruby project I've seen that exposes an AST for MD is https://github.com/phlex-ruby/phlex-markdown, but there are some edge cases with this code like if an ERB block spans multiple lines, it won't evaluate.

**3. Change the ERB tag that evaluates**

Probably the easiest thing to do is change ERB in MD to render with `<%%= render "partial" %%>` instead of `<%= render "partial" %>`. It avoids some of the issues of the AST approach I outlined above.

**4. Other ides**

My ideas to solve this is not the full list! Curious how others tackle this problem. If you've done it a different way please share!

**Comment by alexisbernard** [comment]
@bradgessler I'm curious how you avoid mixing ERB when you need to call Ruby code, for example `asset_path`, from a markdown template ?

By using double extensions the order between MD and ERB becomes obvious : `template.html.md.erb` (ERB then MD) or `template.html.erb.md` (MD then ERB). That means `template.html.md` does not interpret ERB.

The double percent to escape ERB works great : `<%%= 1 + 2 %> => <%= 1 + 2 %>` renders `<%= 1 + 2 %> => 3`.

**Comment by bradgessler** [comment]
Use the `markdown-rails` gem and it will handle all of that for you out of the box with this class: https://github.com/sitepress/markdown-rails/blob/main/lib/markdown-rails/renderer/rails.rb

You can create your own markdown renderer by extending that class to do whatever embeds you want to do.
